File: android/src/main/java/expo/modules/contenttransition/glyphs/GlyphState.kt
package expo.modules.contenttransition.glyphs

import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.RenderEffect
import androidx.compose.ui.text.TextLayoutResult
import expo.modules.contenttransition.animation.Spring
import expo.modules.contenttransition.animation.SpringValue
import expo.modules.contenttransition.animation.TransitionSprings
import expo.modules.contenttransition.records.GlyphPlacement
import expo.modules.contenttransition.records.GlyphSlot
import expo.modules.contenttransition.records.TransitionShape
import expo.modules.contenttransition.records.TypesetLine
import kotlin.math.ceil

internal class GlyphState(
  val id: Long,
  val unit: String,
  springs: TransitionSprings
) {
  var slot: GlyphSlot? = null
  var invalid: Boolean = false
  var disappearing: Boolean = false
  var delay: Float = 0f
  val x = SpringValue(springs.position, 0f, epsilon = 0.05f)
  val offset = SpringValue(springs.offset, 0f, epsilon = 0.002f)
  val scale = SpringValue(springs.glyphScale, 1f, epsilon = 0.002f)
  val alpha = SpringValue(springs.opacity, 1f, epsilon = 0.002f)
  val blur = SpringValue(springs.blurIn, 0f, epsilon = 0.05f)

  private val presenceState = mutableFloatStateOf(1f)
  var presence: Float
    get() = presenceState.floatValue
    set(value) {
      presenceState.floatValue = value
    }

  private val displacedState = derivedStateOf {
    offset.value != 0f || scale.value != 1f || blur.value > 0f
  }
  val isDisplaced: Boolean
    get() = displacedState.value

  var layout: TextLayoutResult? = null
    private set

  var viewportWidth: Int = 1
    private set
  var viewportHeight: Int = 1
    private set

  var drawX: Float = 0f
    private set
  var drawY: Float = 0f
    private set

  var padX: Float = 0f
    private set
  var padY: Float = 0f
    private set

  var travel: Float = 0f
    private set

  var featherMask: Brush? = null
    private set

  private var enterScale: Float = TransitionShape.DEFAULT_ENTER_SCALE
  private var appearBlur: Float = 0f
  private var disappearBlur: Float = 0f
  private var blurIn: Spring = springs.blurIn
  private var blurOut: Spring = springs.blurOut

  fun adoptSprings(springs: TransitionSprings) {
    x.spring = springs.position
    offset.spring = springs.offset
    scale.spring = springs.glyphScale
    alpha.spring = springs.opacity
    blurIn = springs.blurIn
    blurOut = springs.blurOut
    blur.spring = if (disappearing) blurOut else blurIn
  }

  fun updateMetrics(
    placement: GlyphPlacement,
    line: TypesetLine,
    density: Float,
    transition: TransitionShape
  ) {
    layout = placement.layout
    travel = line.height * transition.travelRatio
    enterScale = transition.enterScale

    val ceiling = transition.maxBlurRadius * density
    appearBlur = (line.height * APPEAR_BLUR_RATIO * transition.blurIntensity).coerceAtMost(ceiling)
    disappearBlur = (line.height * DISAPPEAR_BLUR_RATIO * transition.blurIntensity).coerceAtMost(ceiling)

    val feather = ceil(line.height * FEATHER_RATIO)
    padX = ceil(line.height * OVERHANG_RATIO)
    padY = feather

    viewportWidth = ceil(placement.advance + padX * 2f).toInt().coerceAtLeast(1)
    viewportHeight = ceil(line.height + padY * 2f).toInt().coerceAtLeast(1)
    drawX = padX
    drawY = padY + (line.baseline - placement.layout.firstBaseline)
    featherMask = featherBrush(
      height = viewportHeight.toFloat(),
      top = padY,
      bottom = padY + line.height,
      feather = feather
    )
  }

  fun beginAppear(countsDown: Boolean, blurEnabled: Boolean) {
    val from = if (countsDown) -1f else 1f
    offset.reset(from, 0f)
    scale.reset(enterScale, 1f)
    alpha.reset(0f, 1f)
    blur.spring = blurIn
    blur.reset(if (blurEnabled) appearBlur else 0f, 0f)
    disappearing = false
  }

  fun returnToVisible() {
    offset.retarget(0f)
    scale.retarget(1f)
    alpha.retarget(1f)
    blur.spring = blurIn
    blur.retarget(0f)
    disappearing = false
  }

  fun beginDisappear(countsDown: Boolean, blurEnabled: Boolean) {
    val to = if (countsDown) 1f else -1f
    offset.retarget(to)
    scale.retarget(enterScale)
    alpha.retarget(0f)
    blur.spring = blurOut
    blur.retarget(if (blurEnabled) disappearBlur else 0f)
    disappearing = true
  }

  fun snapToVisible() {
    offset.snapTo(0f)
    scale.snapTo(1f)
    alpha.snapTo(1f)
    blur.snapTo(0f)
    disappearing = false
    delay = 0f
  }

  fun clearBlur() {
    blur.snapTo(0f)
  }

  val isAnimating: Boolean
    get() = delay > 0f ||
      !x.isSettled ||
      !offset.isSettled ||
      !scale.isSettled ||
      !alpha.isSettled ||
      !blur.isSettled

  val isVisible: Boolean
    get() = alpha.value >= VISIBILITY_THRESHOLD || alpha.target >= VISIBILITY_THRESHOLD

  fun tick(deltaTime: Float): Boolean {
    var running = false
    if (x.tick(deltaTime)) running = true
    if (offset.tick(deltaTime)) running = true
    if (scale.tick(deltaTime)) running = true
    if (alpha.tick(deltaTime)) running = true
    if (blur.tick(deltaTime)) running = true
    return running
  }

  fun blurEffect(enabled: Boolean, visibleScale: Float = 1f): RenderEffect? =
    if (enabled) GlyphBlur.effect(blur.value * visibleScale) else null

  private companion object {
    const val APPEAR_BLUR_RATIO = 0.25f
    const val DISAPPEAR_BLUR_RATIO = 0.275f
    const val FEATHER_RATIO = 0.35f
    const val OVERHANG_RATIO = 0.15f
    const val FEATHER_STEPS = 6
    const val VISIBILITY_THRESHOLD = 0.01f

    fun featherBrush(height: Float, top: Float, bottom: Float, feather: Float): Brush {
      val stops = ArrayList<Pair<Float, Color>>((FEATHER_STEPS + 1) * 2)
      for (step in 0..FEATHER_STEPS) {
        val t = step / FEATHER_STEPS.toFloat()
        stops.add(((top - feather * (1f - t)) / height) to Color.Black.copy(alpha = smoothstep(t)))
      }
      for (step in 0..FEATHER_STEPS) {
        val t = step / FEATHER_STEPS.toFloat()
        stops.add(((bottom + feather * t) / height) to Color.Black.copy(alpha = smoothstep(1f - t)))
      }

      return Brush.verticalGradient(*stops.toTypedArray(), startY = 0f, endY = height)
    }

    fun smoothstep(t: Float): Float = t * t * (3f - 2f * t)
  }
}

File: web/glyphs/GlyphState.ts
import type { Spring } from '../animation/Spring';
import { SpringValue } from '../animation/SpringValue';
import type { TransitionSprings } from '../animation/TransitionSprings';
import {
  TRANSITION_SHAPE_DEFAULTS,
  type GlyphPlacement,
  type GlyphSlot,
  type TransitionShape,
  type TypesetLine,
} from '../records';
import { GlyphBlur } from './GlyphBlur';

const APPEAR_BLUR_RATIO = 0.25;
const DISAPPEAR_BLUR_RATIO = 0.275;
const FEATHER_RATIO = 0.35;
const OVERHANG_RATIO = 0.15;
const BLUR_EXTENT_SIGMAS = 3;
const VISIBILITY_THRESHOLD = 0.01;

class GlyphState {
  readonly id: number;
  readonly unit: string;

  slot: GlyphSlot | null = null;
  invalid = false;
  disappearing = false;
  delay = 0;

  readonly x: SpringValue;
  readonly offset: SpringValue;
  readonly scale: SpringValue;
  readonly alpha: SpringValue;
  readonly blur: SpringValue;

  presence = 1;

  advance = 0;
  lineHeight = 0;
  travel = 0;
  padX = 0;
  padY = 0;
  feather = 0;
  viewportWidth = 1;
  viewportHeight = 1;

  private enterScale: number = TRANSITION_SHAPE_DEFAULTS.enterScale;
  private appearBlur = 0;
  private disappearBlur = 0;
  private blurIn: Spring;
  private blurOut: Spring;

  constructor(id: number, unit: string, springs: TransitionSprings) {
    this.id = id;
    this.unit = unit;
    this.x = new SpringValue(springs.position, 0, 0.02);
    this.offset = new SpringValue(springs.offset, 0, 0.002);
    this.scale = new SpringValue(springs.glyphScale, 1, 0.002);
    this.alpha = new SpringValue(springs.opacity, 1, 0.002);
    this.blur = new SpringValue(springs.blurIn, 0, 0.05);
    this.blurIn = springs.blurIn;
    this.blurOut = springs.blurOut;
  }

  adoptSprings(springs: TransitionSprings): void {
    this.x.spring = springs.position;
    this.offset.spring = springs.offset;
    this.scale.spring = springs.glyphScale;
    this.alpha.spring = springs.opacity;
    this.blurIn = springs.blurIn;
    this.blurOut = springs.blurOut;
    this.blur.spring = this.disappearing ? this.blurOut : this.blurIn;
  }

  updateMetrics(placement: GlyphPlacement, line: TypesetLine, shape: TransitionShape): void {
    this.advance = placement.advance;
    this.lineHeight = line.height;
    this.travel = line.height * shape.travelRatio;
    this.enterScale = shape.enterScale;

    this.appearBlur = Math.min(
      line.height * APPEAR_BLUR_RATIO * shape.blurIntensity,
      shape.maxBlurRadius
    );
    this.disappearBlur = Math.min(
      line.height * DISAPPEAR_BLUR_RATIO * shape.blurIntensity,
      shape.maxBlurRadius
    );

    const blurExtent = Math.ceil(
      GlyphBlur.sigma(Math.max(this.appearBlur, this.disappearBlur)) * BLUR_EXTENT_SIGMAS
    );

    this.feather = Math.ceil(line.height * FEATHER_RATIO);
    this.padX = Math.max(Math.ceil(line.height * OVERHANG_RATIO), blurExtent);
    this.padY = this.feather;

    this.viewportWidth = Math.max(Math.ceil(placement.advance + this.padX * 2), 1);
    this.viewportHeight = Math.max(Math.ceil(line.height + this.padY * 2), 1);
  }

  beginAppear(countsDown: boolean, blurEnabled: boolean): void {
    const from = countsDown ? -1 : 1;
    this.offset.reset(from, 0);
    this.scale.reset(this.enterScale, 1);
    this.alpha.reset(0, 1);
    this.blur.spring = this.blurIn;
    this.blur.reset(blurEnabled ? this.appearBlur : 0, 0);
    this.disappearing = false;
  }

  returnToVisible(): void {
    this.offset.retarget(0);
    this.scale.retarget(1);
    this.alpha.retarget(1);
    this.blur.spring = this.blurIn;
    this.blur.retarget(0);
    this.disappearing = false;
  }

  beginDisappear(countsDown: boolean, blurEnabled: boolean): void {
    const to = countsDown ? 1 : -1;
    this.offset.retarget(to);
    this.scale.retarget(this.enterScale);
    this.alpha.retarget(0);
    this.blur.spring = this.blurOut;
    this.blur.retarget(blurEnabled ? this.disappearBlur : 0);
    this.disappearing = true;
  }

  snapToVisible(): void {
    this.offset.snapTo(0);
    this.scale.snapTo(1);
    this.alpha.snapTo(1);
    this.blur.snapTo(0);
    this.disappearing = false;
    this.delay = 0;
  }

  clearBlur(): void {
    this.blur.snapTo(0);
  }

  get isDisplaced(): boolean {
    return this.offset.value !== 0 || this.scale.value !== 1 || this.blur.value > 0;
  }

  get isAnimating(): boolean {
    return (
      this.delay > 0 ||
      !this.x.isSettled ||
      !this.offset.isSettled ||
      !this.scale.isSettled ||
      !this.alpha.isSettled ||
      !this.blur.isSettled
    );
  }

  get isVisible(): boolean {
    return this.alpha.value >= VISIBILITY_THRESHOLD || this.alpha.target >= VISIBILITY_THRESHOLD;
  }

  tick(deltaTime: number): boolean {
    let running = false;
    if (this.x.tick(deltaTime)) running = true;
    if (this.offset.tick(deltaTime)) running = true;
    if (this.scale.tick(deltaTime)) running = true;
    if (this.alpha.tick(deltaTime)) running = true;
    if (this.blur.tick(deltaTime)) running = true;
    return running;
  }
}

export { GlyphState };
